src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs

```diff
@@ -1556,6 +1556,24 @@ internal CommonAssemblyWellKnownAttributeData GetSourceDecodedWellKnownAttribute
             return (CommonAssemblyWellKnownAttributeData)attributesBag.DecodedWellKnownAttributeData;
         }
 
+        // This only forces binding of attributes that look like they may be forwarded types attributes (syntactically).
+        internal HashSet<NamedTypeSymbol> GetForwardedTypes()
+        {
+            CustomAttributesBag<CSharpAttributeData> attributeBag = null;
+            LoadAndValidateAttributes(OneOrMany.Create(GetAttributeDeclarations()), ref attributeBag, attributeMatchesOpt: this.IsForwardedTypesAttribute);
+
+            var wellKnownAttributeData = (CommonAssemblyWellKnownAttributeData)attributeBag?.DecodedWellKnownAttributeData;
+            return wellKnownAttributeData?.ForwardedTypes;
+        }
+
+        private bool IsForwardedTypesAttribute(AttributeSyntax node)
+        {
+            QuickAttributeChecker checker =
+                this.DeclaringCompilation.GetBinderFactory(node.SyntaxTree).GetBinder(node).QuickAttributeChecker;
```

Comment:
>this.DeclaringCompilation.GetBinderFactory(node.SyntaxTree).GetBinder(node).QuickAttributeChecker [](start = 16, length = 97)

It is unfortunate that we are doing this for each attribute, the checker could be shared between all attributes in the same list.

Reply:
Yes, it would be nice to re-use the checker. But each attribute could be in a different file/syntaxTree with different usings.

---
In reply to: [155961077](https://github.com/dotnet/roslyn/pull/23647#discussion_r155961077) [](ancestors = 155961077)

Reply:
> But each attribute could be in a different file/syntaxTree with different usings.

Syntactically attributes are grouped in lists, all attributes from the same list are using the same context.

---
In reply to: [156193298](https://github.com/dotnet/roslyn/pull/23647#discussion_r156193298) [](ancestors = 156193298,155961077)